source/server/session_repository.cpp

```diff
@@ -2,6 +2,13 @@
 
 #include <grpcpp/grpcpp.h>
 
+#if !defined(NI_SESSION_REQUESTED_NEW_SESSION_BUT_ONE_ALREADY_EXISTS_ERROR)
```

Comment:
These are returned as driver error codes, right? I think it would make sense to throw an exception that is translated to gRPC status code `NOT_FOUND` or `ALREADY_EXISTS`, similar to https://github.com/ni/grpc-device/pull/687 . Returning a driver error code is more appropriate for driver errors.

![image](https://user-images.githubusercontent.com/17914399/195197376-f2d29b1b-bf3b-4227-b5c3-233d39763165.png)
![image](https://user-images.githubusercontent.com/17914399/195197444-24c946a0-019d-497e-9956-75f387f9948c.png)

Reply:
I hope it was obvious that -1 and -2 were placeholders. That's why I put [RFC] in the title. I guess I could've made it a draft, but to me that says "I'm still working on this, don't comment yet" which is the opposite of what I wanted.

I would almost say yes to `ALREADY_EXISTS` but I can't agree to `NOT_FOUND` since we return that when the library's method pointer is NULL, which would usually indicate a newer gRPC-device server with an older driver runtime.

Reply:
What about `FAILED_PRECONDITION` instead of `NOT_FOUND`? I could almost argue that would be a meaningful status code for both cases. But would that be stretching it?

*The operation was rejected because the system is not in a state required for the operation's execution. For example, the directory to be deleted is non-empty, an rmdir operation is applied to a non-directory, etc. Service implementors can use the following guidelines to decide between FAILED_PRECONDITION, ABORTED, and UNAVAILABLE: (a) Use UNAVAILABLE if the client can retry just the failing call. (b) Use ABORTED if the client should retry at a higher level (e.g., when a client-specified test-and-set fails, indicating the client should restart a read-modify-write sequence). (c) Use FAILED_PRECONDITION if the client should not retry until the system state has been explicitly fixed. E.g., if an "rmdir" fails because the directory is non-empty, FAILED_PRECONDITION should be returned since the client should not retry unless the files are deleted from the directory.*

Reply:
> I hope it was obvious that -1 and -2 were placeholders.

Yes. What is it a placeholder for? An infrastructure error like `niapalerr_resourceNotFound`? Are you planning to add an error code range for grpc-device specific errors? 

Anyway, this code used to return +1. See https://github.com/ni/grpc-device/issues/685 and https://github.com/ni/grpc-device/pull/687 . That was before you changed grpc-device to return errors as UNKNOWN, so the example programs would ask the driver to look up the error message for +1.

> I would almost say yes to ALREADY_EXISTS but I can't agree to NOT_FOUND since we return that when the library's method pointer is NULL, which would usually indicate a newer gRPC-device server with an older driver runtime.

The server returns NOT_FOUND when the operation is UNIMPLEMENTED by the driver? 😅

> What about FAILED_PRECONDITION instead of NOT_FOUND? I could almost argue that would be a meaningful status code for both cases. But would that be stretching it?

FAILED_PRECONDITION is plausible but ALREADY_EXISTS and NOT_FOUND seem clearer to me. The session/task is an entity that either already exists or was not found.

Note that gRPC status codes also include a developer-facing error message which can explain what was not found.

Reply:
Given our existing use of NOT_FOUND, I chose FAILED_PRECONDITION and ALREADY_EXISTS for this PR. I'm not opposed to those changing in the future, but it will be easier to make that change now that the gRPC status code is tied to what is thrown rather than what is caught.